# Linear Regression
The Whopper has been Burger King’s signature sandwich since 1957. One Double Whopper with cheese provides 53 grams of protein! That's all the protein you need in a day. It also supplies 1020 calories and 65 grams of fat 😲😲😲. FYI - the Daily Value (based on a 2000-calorie diet) for fat is 65 grams. So after a Double Whopper you’ll want the rest of your calories that day to be fat-free. Of course, the Whopper isn’t the only item Burger King sells. How are fat and protein related on the entire BK menu? The scatterplot of the Fat (in grams) versus the Protein (in grams) for foods sold at Burger King shows a positive, moderately strong, linear relationship.
![Scatterplot of BK](https://raw.githubusercontent.com/SSpindt/AI/refs/heads/main/Semester%201/Problem%20Set%20Images/BKScatter.png)
If you want 25 grams of protein in your lunch, how much fat should you expect to consume at Burger King? The correlation between Fat and Protein is 0.83, a sign that the linear association seen in the scatterplot is fairly strong. But strength of the relationship is only part of the picture. The correlation says, “The linear association between these two variables is fairly strong,” but it doesn’t tell us what the line actually is.
Now we can say more; the equation will let us predict the fat content for any Burger King food, given its amount of protein. 

For the Burger King foods, we’d choose a linear model to describe the relationship between Protein and Fat.
Of course, no line can go through all the points, but a linear model can summarize the general pattern with only a couple of parameters. Like all models of the real world, the line will be wrong — wrong in the sense that it can’t match reality exactly. But it can help us understand how the variables are associated.
Not only can’t we draw a line through all the points, the best line might not even hit any of the points. Then how can it be the “best” line? We want to find the line that somehow comes closer to all the points than any other line. Some of the points will be above the line and some below. For example, the line might suggest that a BK Broiler chicken sandwich with 30 grams of protein should have 36 grams of fat when, in fact, it actually has only 25 grams of fat. The BK Broiler chicken residual would be `actual - predicted = 25 - 36 = -11 g of fat`. The negative residual tells us that the actual fat content of the BK Broiler chicken is about 11 grams less than the model predicts for a typical Burger King menu item with 30 grams of protein. Our challenge now is how to find the right line.

When we draw a line through a scatterplot, some residuals are positive and some negative. We can’t assess how well the line fits by adding up all the residuals—the positive and negative ones would just cancel each other out. We faced the same issue when we calculated a standard deviation to measure spread. And we deal with it the same way here: by squaring the residuals. Squaring makes them all positive. Now we can add them up. Squaring also emphasizes the large residuals. After all, points near the line are consistent with the model, but we’re more concerned about points far from the line. When we add all the squared residuals together, that sum indicates how well the line we drew fits the data — the smaller the sum, the better the fit. A different line will produce a different sum, maybe bigger, maybe smaller. The **line of best fit** is the line for which the sum of squared residuals is smallest! We call this line the **least squares line**. It can be described as `y = mx + b `or more generically as `y = b0 + b1x` (*b0 is the y intercept and b1 is the slope*).You can use the line of best fit to make predictions -- all you have to do is plug in your value for x and solve for y!

## <span style="color:blue">Your Turn</span>
Please solve the following problems using Python, Markdown, and this [data set about movies](https://drive.google.com/file/d/1QmsmZSTI9VOesBH4yL_8LHTxV63Gz2hv/view?usp=drive_link). 

## <span style="color:blue">Problem 1</span>
First, examine the relationship between `budget` and `critics score`. Then, answer these questions:
1. What kind of relationship is found between `budget` and `critics score`? How do you know?
2. Define a function that returns the equation of the line of best fit for `budget` and `critics score`. What is that equation? *You can only use standard Python and Pandas in your function. You cannot use Numpy or any other Python library.*
3. According to your line of best fit, what score would a movie receive if its budget was $10 million? How accurate is this prediction?

In [4]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
df = pd.read_json('problem_set.json')
print(df.dtypes)
print(df['budget (millions)'].corr(df['critics score (percent)']))

critics score (percent)             float64
budget (millions)                   float64
length (minutes)                    float64
curse words used (rate per hour)    float64
dtype: object
0.9697766428756551


The two variables have a strong relationship, as the corelation coeffecent 0.9697 is very close to 1, indicicating a strong positive relationship.

In [18]:
def best_fit_equation(x_col, y_col):
    x_mean = df[x_col].mean()
    y_mean = df[y_col].mean()

    m = ((df[x_col] * df[y_col]).mean() - x_mean * y_mean) / ((df[x_col]**2).mean() - x_mean**2)

    b = y_mean - m * x_mean
    
    return m, b

best_fit_equation('budget (millions)','critics score (percent)')

(0.46603264423232343, -1.3987314019555512)

The equation for a line of best fit is "y = 0.47x - 1.40".

In [21]:
def critics_score(budget, x_col, y_col):
    m, b = best_fit_equation(x_col, y_col)
    return m * budget + b

critics_score(10, 'budget (millions)', 'critics score (percent)')

3.261595040367683

Acording to our best fit line a movie with a budget of 10m would get a score of 3.3%. This is inacurate, and is likely because the score to budget ratio is non-linear in reality.

## <span style="color:blue">Problem 2</span>
Now, examine the relationship between `length` and `critics score`. Then, answer these questions:
1. What kind of relationship is found between `length` and `critics score`? How do you know?
2. If you haven't done so already, please generalize your Linear Regression function so that it works on all kinds of quantitative data. Once you are done generalizing your function, call it using `length` and `critics score` as arguments. What is the equation of the best fit line? *Reminder, you can only use standard Python and Pandas in your function. You cannot use Numpy or any other Python library.*
3. According to your line of best fit, what score would a movie receive if its length was 150 minutes? How accurate is this prediction?

In [24]:
print(df['length (minutes)'].corr(df['critics score (percent)']))

-0.7649508928065849


There's a moderately strong negitive relationship, meaning longer movies often get lower scores. This is known because .-765 is close to -1.

In [26]:
best_fit_equation('length (minutes)','critics score (percent)')

(-0.6341943700384406, 132.6733124586916)

The equation for a line of best fit is "y = -0.63x + 132.67".

In [31]:
critics_score(150, 'length (minutes)', 'critics score (percent)')

37.54415695292552

Decently accurate, Rotten Tomatos has an average score of 62, so 38%'s not that far off

## <span style="color:blue">Problem 3</span>
Finally, examine the relationship between `budget` and `curse words used`. Then, answer these questions:
1. What kind of relationship is found between `budget` and `curse words used`? How do you know?
2. Generalize your Linear Regression function even more. You should be able to call it and have it work on any quantitative data without changing anything about the function definition. Once you are done generalizing your function, call it using `budget` and `curse words used` as arguments. What is the equation of the best fit line? *Reminder, you can only use standard Python and Pandas in your function. You cannot use Numpy or any other Python library.*
3. According to your line of best fit, how many curse words would be used during a movie if it had a $5 million dollar budget? How accurate is this prediction?

In [34]:
print(df['budget (millions)'].corr(df['curse words used (rate per hour)']))

0.035462236713153815


There is little tobest_fit_equation('length (minutes)','critics score (percent)') no corrilation between the number of curse words per hour and the budget of a film.

In [37]:
best_fit_equation('budget (millions)','curse words used (rate per hour)')

(0.00578065095976991, 9.240233598815554)

y = .005x + 9.240

In [39]:
critics_score(5, 'budget (millions)', 'curse words used (rate per hour)')

9.269136853614404

About 9 curse words per hour. This seems to line up with most data on the average amount of curse words in movies, but it doesn't seem to be related to the budget

## <span style="color:blue">Problem 4</span>
Compare Linear Regression algorithms with someone from a different table. Then, answer these questions:
1. How are your algorithms similar? Different?
2. Whose algorithm was more accurate for problem 1? Problem 2? Problem 3?
3. Why do you think that algorithm was more accurate for one or all of these problems?

1. I'm doing these questions on my own at home, so it's impossible for me to answer question 4.  If someone wanted to, they could use a library like nunpy to have built in versions of a lot of these functions. My code differs by only using Python and Pandas

2. Nunpy would be the most accurate for each question, which is why it's an industry standard

3. Nunpy's the most accurate because it needs to be used by tons of people each day. It also makes efforts to only be as large as it needs to be for this reason.